examples/ListObjects.java

```diff
@@ -65,6 +65,18 @@ public static void main(String[] args)
         }
       }
 
+     {
+        // Listing object information contains usermetadata.
+        Iterable<Result<Item>> results =
+            minioClient.listObjects(
+                ListObjectsArgs.builder().bucket("my-bucketname").includeUserMetadata(true).build());
```

Comment:
`includeUserMetadata` is only applicable to MinIO server hence this example is not needed

Reply:
Ok, I encountered this problem when using minio server java sdk. The related introduction of listObjects method is very short.

Therefore, I suggest that you can add relevant introduction or example

The link I used is
`https://min.io/docs/minio/linux/developers/java/API.html#listobjects-listobjectsargs-args`

And the documentation doesn't say anything about "update object metadata"
My current implementation is:

1 stat_object ->get old object meta
2 old object meta ->add new-usermetadata，create new object meta
3 copy old object -> other ojbect

I don't think it's good enough. Is there any other better way

Reply:
thanks

Reply:
Yep. You have to copy object with new/updated metadata. Here destination object can be the same.

Reply:
I saw this post
#https://github.com/minio/minio/pull/3487
so,I implemented the object'metadata update as described API
`
minioClient.copyObject(
  CopyObjectArgs.builder()
    .bucket(bucketName)
    .object(filePath)
    .userMetadata(newUserMate)
    .source(
        CopySource.builder()
            .bucket(bucketName)
            .object(filePath)
            .build()
    ).build()
`
newUserMate is Map object, it is not equal to oldUserMate
oldUserMate:{"key":"oldValue"}  -> newUserMate :{"key":"newVal"}

but, l run into a error:
`
This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes
`

This error is similar to the post linked below:
[https://github.com/minio/minio/issues/3316](url)

Excuse me, where do I use the error

thanks for your reply!!

Reply:
Enable `MinioClient.traceOn(System.out)`  and see what's going on

Reply:
error log:
---------START-HTTP---------
PUT /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue HTTP/1.1
x-amz-copy-source: /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue
X-Amz-Meta-Filename: dabssssichi.vue
Host: 14.116.220.146:9000
Accept-Encoding: identity
User-Agent: MinIO (amd64; amd64) minio-java/3.19.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20230202T081943Z
Authorization: ██

HTTP/1.1 400
Accept-Ranges: bytes
Content-Length: 543
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO
Vary: Origin
X-Amz-Request-Id: 173FF415CCDDA832
X-Xss-Protection: 1; mode=block
Date: Thu, 02 Feb 2023 08:19:20 GMT

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidRequest</Code><Message>This copy request is illegal because it is trying to copy an object to itself without changing the object&#39;s metadata, storage class, website redirect location or encryption attributes.</Message><Key>2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue</Key><BucketName>jrsy</BucketName><Resource>/jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue</Resource><RequestId>173FF415CCDDA832</RequestId><HostId>eb85c247-f8ab-4fd3-a17a-fdc628cc014b</HostId></Error>
----------END-HTTP----------
2023-02-02 16:19:43.945 ERROR 10664 --- [nio-8800-exec-2] c.j.e.c.f.service.impl.MinioServiceImpl  : This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.
error occurred
ErrorResponse(code = InvalidRequest, message = This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes., bucketName = jrsy, objectName = 2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue, resource = /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue, requestId = 173FF415CCDDA832, hostId = eb85c247-f8ab-4fd3-a17a-fdc628cc014b)
request={method=PUT, url=http://14.116.220.146:9000/jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue, headers=x-amz-copy-source: /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue
X-Amz-Meta-Filename: dabssssichi.vue
Host: 14.116.220.146:9000
Accept-Encoding: identity
User-Agent: MinIO (amd64; amd64) minio-java/3.19.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20230202T081943Z
Authorization: ██
}
response={code=400, headers=Accept-Ranges: bytes
Content-Length: 543
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO
Vary: Origin
X-Amz-Request-Id: 173FF415CCDDA832
X-Xss-Protection: 1; mode=block
Date: Thu, 02 Feb 2023 08:19:20 GMT
}

	at io.minio.MinioClient.execute(MinioClient.java:1135)
	at io.minio.MinioClient.execute(MinioClient.java:935)
	at io.minio.MinioClient.executePut(MinioClient.java:1252)
	at io.minio.MinioClient.copyObject(MinioClient.java:2081)
	at com.jrsy.erp.common.file.service.impl.MinioServiceImpl.updateFileForUid(MinioServiceImpl.java:313)
	at com.jrsy.erp.common.file.controller.FileController.forUidUpdate(FileController.java:114)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:920)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:684)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1743)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)

Reply:
**HTTP LOG**
---------START-HTTP---------
PUT /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue HTTP/1.1
x-amz-copy-source: /jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue
X-Amz-Meta-Filename: dabssssichi.vue
Host: xxx.xxx.xxx.xxx:9000
Accept-Encoding: identity
User-Agent: MinIO (amd64; amd64) minio-java/3.19.0
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20230202T081943Z
Authorization: ██

HTTP/1.1 400
Accept-Ranges: bytes
Content-Length: 543
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO
Vary: Origin
X-Amz-Request-Id: 173FF415CCDDA832
X-Xss-Protection: 1; mode=block
Date: Thu, 02 Feb 2023 08:19:20 GMT

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidRequest</Code><Message>This copy request is illegal because it is trying to copy an object to itself without changing the object&#39;s metadata, storage class, website redirect location or encryption attributes.</Message><Key>2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue</Key><BucketName>jrsy</BucketName>
ource>/jrsy/2/index.vue.76f119fe7fff4b7184f87873ad788fbb.vue</Resource
<RequestId>173FF415CCDDA832</RequestId><HostId>eb85c247-f8ab-4fd3-a17a-fdc628cc014b</HostId></Error>
----------END-HTTP----------

Error given by MINIO server log
The info core is "This copy request is illegal because it is trying to copy an object to itself without changing the object&#39;s metadata, storage class, website redirect location or encryption attributes."

Reply:
Refer https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html and https://minio-java.min.io/io/minio/CopyObjectArgs.html

Reply:
It is explained in https://github.com/minio/minio/pull/3487
`
This is written so that to simplify our handler code
and provide a way to only update metadata instead of
the data when source and destination in CopyObject
request are same.
`
so,l think use "copyObject", Go directly to modify file'usermetadata

![image](https://user-images.githubusercontent.com/72176072/216273341-cd71d6ac-1557-4eac-9867-939412f43b5a.png)

Only usermetadata's data is updated

Reply:
It is explained in https://github.com/minio/minio/pull/3487


> 

ok